pkg/ddl/schema.go

```diff
@@ -154,6 +161,144 @@ func onModifySchemaDefaultPlacement(jobCtx *jobContext, job *model.Job) (ver int
 	return ver, nil
 }
 
+func (w *worker) onModifySchemaReadOnly(jobCtx *jobContext, job *model.Job) (ver int64, err error) {
+	args, err := model.GetModifySchemaArgs(job)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	dbInfo, err := checkSchemaExistAndCancelNotExistJob(jobCtx.metaMut, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	if job.SchemaState == model.StateNone && dbInfo.ReadOnly == args.ReadOnly {
+		job.FinishDBJob(model.JobStateDone, model.StatePublic, ver, dbInfo)
+		return ver, nil
+	}
+	// If the database is set to read-write from read-only, we don't need the middle state.
+	if dbInfo.ReadOnly && !args.ReadOnly {
+		dbInfo.ReadOnly = args.ReadOnly
+		if err = jobCtx.metaMut.UpdateDatabase(dbInfo); err != nil {
+			return ver, errors.Trace(err)
+		}
+		if ver, err = updateSchemaVersion(jobCtx, job); err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.FinishDBJob(model.JobStateDone, model.StatePublic, ver, dbInfo)
+		return ver, nil
+	}
+	trxTableName := infoschema.ClusterTableTiDBTrx
+	failpoint.Inject("mockModifySchemaReadOnlyDDL", func() {
+		trxTableName = infoschema.TableTiDBTrx
+	})
+
+	switch job.SchemaState {
+	case model.StateNone:
+		dbInfo.ReadOnly = args.ReadOnly
+		if err = jobCtx.metaMut.UpdateDatabase(dbInfo); err != nil {
+			return ver, errors.Trace(err)
+		}
+		if ver, err = updateSchemaVersion(jobCtx, job); err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StatePendingReadOnly
+	case model.StatePendingReadOnly:
+		uncommittedTxn, err := getUncommittedTxnIDs(jobCtx, w.sess, job.Query, dbInfo.ID, trxTableName)
+		failpoint.Inject("mockCheckUncommittedTxnError", func() {
+			if err == nil {
+				err = errors.New("mock error for check uncommitted txn")
+			}
+		})
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// check and wait all uncommitted txn
+		for len(uncommittedTxn) > 0 {
+			for txnID := range uncommittedTxn {
+				sql := fmt.Sprintf("SELECT ID FROM INFORMATION_SCHEMA.%s WHERE ID = %d", trxTableName, txnID)
+				r, err := w.sess.Execute(jobCtx.stepCtx, sql, "check if txn committed")
+				if err != nil {
+					return ver, errors.Trace(err)
+				}
+				if len(r) == 0 {
+					delete(uncommittedTxn, txnID)
+					continue
+				}
+				logutil.BgLogger().Info("uncommitted txn block read only ddl", zap.Int64("txn ID", txnID))
```

Comment:
Maybe use `SampleLogger` instead to prevent too many logs. Also add the job ID to zap fields.

Reply:
I tried the `SampleLogger` and found the different message will not be output, the `id=2` log is gone.  Is there something wrong with my usage?
```
func TestLog(t *testing.T) {
	sampleLogger := logutil2.SampleLoggerFactory(
		time.Minute, 3, zap.String(logutil2.LogFieldCategory, "ddl"),
	)

	sampleLogger().Info("test log", zap.Int("id", 1))
	sampleLogger().Info("test log", zap.Int("id", 1))
	sampleLogger().Info("test log", zap.Int("id", 1))
	sampleLogger().Info("test log", zap.Int("id", 2))
	sampleLogger().Info("test log", zap.Int("id", 2))
}
```

Reply:
Besides this comment, all fixed in https://github.com/pingcap/tidb/pull/62804/commits/3ca980016ce8c7087c974afa084fe639f4810b7f, PTAL again